libs/platform/core/src/services/query/core/core-query.ts

```diff
@@ -0,0 +1,233 @@
+import {
+  BehaviorSubject,
+  catchError,
+  distinctUntilChanged,
+  filter,
+  from,
+  map,
+  merge,
+  NEVER,
+  Observable,
+  shareReplay,
+  Subscription,
+  switchMap,
+  take,
+  takeUntil,
+  tap,
+  using,
+} from 'rxjs';
+import { CoreQueryService } from '../core';
+import {
+  Query,
+  QueryEventHandler,
+  QueryOptions,
+  QueryState,
+  QueryTrigger,
+} from '../models';
+import { buildEvent } from './build-event';
+
+export class CoreQuery<
+  ValueType,
+  Qualifier extends object | undefined = undefined
+> implements Query<ValueType, Qualifier>
+{
+  protected queryCache = new Map<
+    string,
+    {
+      subject$: BehaviorSubject<QueryState<ValueType>>;
+      state$: Observable<QueryState<ValueType>>;
+      data$: Observable<ValueType | undefined>;
+    }
+  >();
+
+  constructor(
+    protected options: QueryOptions<ValueType, Qualifier>,
+    protected service: CoreQueryService,
+    protected destroyNotifier$?: Observable<undefined>
+  ) {}
+
+  protected getKey(qualifier: Qualifier): string {
+    return qualifier
+      ? JSON.stringify(
+          Object.entries(qualifier)
+            .filter((x) => x[1])
+            .sort()
+            .flat()
```

Comment:
mb we should change order?

Reply:
no, we want to sort parameters of qualifier before faltting values. Flatting it uset to make key string more compact.